File: editor.v2/bsp_tree.py
from editor_math import *
from map import *

SUBSECTOR = 0x8000
MAX_INT = 2 ** 32

class DivLine:
    def __init__(self, vert, dx, dy):
        self.vert = vert
        self.dx = dx
        self.dy = dy

class BSP_Node:
    def __init__(self, id, wall_id, front, back,  walls):
        self.id = id
        self.front = front
        self.back = back
        self.wall_id = wall_id if wall_id is not None else -1
        self.walls = walls

class Subsector:
    def __init__(self, node_id, walls):
        self.node_id = node_id
        self.walls = walls

class BSP_Tree:
    def __init__(self, map):
        self.ref_map = map
        self.bsp_head = None
        self.max_id = 0

        # Generate BSP Tree
        self.generate_bsp_tree()

    def generate_bsp_tree(self):
        # Set bsp_head to the 
        self.subsectors = []
        self.bsp_head = self.build_bsp_node(self.ref_map.Walls)

    def split_wall(self, div_line, wall):
        intersection = self.get_intersection(div_line, wall.line)

        new_vert = self.ref_map.AddVertex(intersection.x, intersection.y)

        v2_id = wall.line.v2.id
        wall.line.v2 = new_vert

        wall2 = self.ref_map.AddWall(
            new_vert.id, v2_id, 
            wall.color, 
            wall.min_height, wall.max_height, 
            wall.floor_height, wall.ceiling_height)

        # Return in front, back order
        if self.line_on_side(div_line, wall.line) == 0:
            return wall, wall2
        else:
            return wall2, wall

    def evaluate_split(self, split_wall, walls, best_grade):
        split_line = self.get_div_line_from_line(split_wall.line)

        front_count, back_count = 0,0

        for w in walls:
            if w.id == split_wall.id:
                continue
            else:
                side = self.line_on_side(split_line, w.line)
            
            if side == 0:
                front_count += 1
            elif side == 1:
                back_count += 1
            elif side != -1:
                front_count += 1
                back_count += 1

            max_count = max(front_count, back_count)
            new_count = front_count + back_count - len(walls)
            grade = max_count + new_count * 8 # ??? This comes from DoomBSP, why the 8?

            # Get out early, no point in continuing
            if grade > best_grade:
                return grade
        if front_count == 0 or back_count == 0:
            return MAX_INT
        
        return grade

    def pick_split_wall(self, walls):
        best_score = MAX_INT
        best_wall = None

        for w in walls:
            score = self.evaluate_split(w, walls, best_score)
            if score < best_score:
                best_score = score
                best_wall = w
        return best_wall, best_score

    def execute_split(self, split_wall, walls):
        split_line = DivLine(split_wall.line.v1, 
                                split_wall.line.v1.x - split_wall.line.v2.x, 
                                split_wall.line.v1.y - split_wall.line.v2.y)

        front_walls = []
        back_walls = []

        for i, w in enumerate(walls):
            if w.id == split_wall.id:
                side = -1
            else:
                side = self.line_on_side(split_line, w.line)

            if side == -1:  # Coincident walls should probably go in both front and back
                front_walls.append(w)
                back_walls.append(w)
            elif side == 0:
                front_walls.append(w)
            elif side == 1:
                back_walls.append(w)
            else:
                wall1, wall2 = self.split_wall(split_line, w)
                front_walls.append(wall1)
                back_walls.append(wall2)
        
        return front_walls, back_walls
        
    # -1 => Coincident
    # -2 => Intersecting
    #  0 => Front
    #  1 => Behind
    def line_on_side(self, div_line, line):
        v1_side = self.vert_on_side(div_line, line.v1)
        v2_side = self.vert_on_side(div_line, line.v2)

        # Coincident
        if v1_side == -2 and v2_side == -2:
            return -1
        
        # Both on same side
        elif v1_side == v2_side:
            return v1_side
        
        # Intersecting at an edge
        elif v1_side == -2:
            return v2_side
        elif v2_side == -2:
            return v1_side

        # Intersecting in the center
        else:
            return -2

    # -2 => Point lies on line
    #  0 => Point is in front of line
    #  1 => Point lies behind line
    def vert_on_side(self, div_line, v):
        # Get normal vector of div_line
        n_dx, n_dy = -1 * div_line.dy, div_line.dx

        # Get vector from div_line.vert to v
        dv_dx, dv_dy = div_line.vert.x - v.x, div_line.vert.y - v.y

        dot = dv_dx * n_dx + dv_dy * n_dy

        if dot == 0:
            return -2
        elif dot > 0:
            return 1
        else:
            return 0

    def get_intersection(self, div_line, line):
        # Solving Ax = B for x => x = A^-1 B
        A = Matrix2x2(  -1 * div_line.dx, line.v2.x - line.v1.x,
                        -1 * div_line.dy, line.v2.y - line.v1.y)
        
        B = Vect2(  div_line.vert.x - line.v1.x,
                    div_line.vert.y - line.v1.y)

        # If det(A) == 0, lines are coincident, so just return v1
        if Det(A) == 0:
            return line.v1 

        x = Mult(Inverse(A), B) # This x vector contains our [t, s]
        t = x.a

        # Calculate intersection based on parameterized line
        intersect = Vertex(-1,  div_line.vert.x + (div_line.dx * t), 
                                div_line.vert.y + (div_line.dy * t))

        return intersect

    def get_div_line_from_line(self, line):
        return DivLine(line.v1, 
                        line.v2.x - line.v1.x, 
                        line.v2.y - line.v1.y)

    def build_bsp_node(self, walls):

        # Get the split wall
        split_wall, wall_score = self.pick_split_wall(walls)

        # Base case: wall_score is MAX_INT => no wall is suitable for partitioning => set of walls is convex
        if wall_score == MAX_INT:
            self.subsectors.append(Subsector(self._get_node_id() + SUBSECTOR, walls))
            return BSP_Node(self.subsectors[-1].node_id, None, None, None, walls)

        front_walls, back_walls = self.execute_split(split_wall, walls)

        return_node = BSP_Node(self._get_node_id(), split_wall.id, None, None, [])
        return_node.front = self.build_bsp_node(front_walls)
        return_node.back = self.build_bsp_node(back_walls)

        return return_node

    # Return structure:
    # (id=int, front=(), back=()), # is a special character denoting an empty child
    def to_string(self):
        return self._to_string_helper(self.bsp_head)

    def _to_string_helper(self, cur_head):
        if cur_head is None:
            return "#"

        front_string = self._to_string_helper(cur_head.front)
        back_string = self._to_string_helper(cur_head.back)

        return "(id=" + str(cur_head.id) + ",wall_id=" + str(cur_head.wall_id) + ",front=" + front_string + ",back=" + back_string + ")"

    def get_map(self):
        return self.ref_map

    def _get_node_id(self):
        self.max_id += 1
        return self.max_id

    # Return a list of sectors in the bsp tree
    def get_sectors(self):
        return self.subsectors

File: editor.v3/com/models/BSPTree.java
package com.models;

import java.util.ArrayList;

public class BSPTree 
{
    public static int MAX_INT = (int)Math.pow(2, 32);
    public static int SUBSECTOR = 0x8000;

    public MapDat Map;
    public BSPNode Head;
    public ArrayList<BSPNode> Subsectors = new ArrayList<BSPNode>();
    private int NextNodeId = 0;
    private int NextWallId = -1;

    private enum Side
    {
        UNDEFINED,
        COINCIDENT,
        INTERSECTING,
        FRONT,
        BACK
    }

    private class DivLine
    {
        public Pair<Float, Float> Position;
        public Pair<Float, Float> Direction;
    }

    public class BSPNode
    {
        public int Id;
        public Pair<BSPNode, BSPNode> Sides;
        public Wall PartitionWall;
        public ArrayList<Wall> Walls;
        public BSPNode() { }
    }

    public BSPTree() { }

    public BSPTree(MapDat map)
    {
        this.Map = map;
        this.Build();
    } 

    public void Build()
    {
        ArrayList<Wall> walls = new ArrayList<Wall>(this.Map.Walls);
        this.Head = BuildNode(walls);
    }

    private BSPNode BuildNode(ArrayList<Wall> walls)
    {
        // Get the splitting wall
        Pair<Wall, Integer> Splitter = PickSplitWall(walls);
        Wall splitWall = Splitter.a;
        int score = Splitter.b;

        BSPNode node = new BSPNode();

        // Base case - the set of walls passed forms a subsector
        if (score == MAX_INT)
        {
            node.Id = this.GetNodeId() + SUBSECTOR;
            node.Walls = walls;
            node.PartitionWall = walls.get(0);

            this.Subsectors.add(node);
            return node;
        }

        Pair<ArrayList<Wall>, ArrayList<Wall>> split = ExecuteSplit(splitWall, walls);

        node.Id = this.GetNodeId();
        node.PartitionWall = splitWall;
        node.Sides = new Pair<BSPNode, BSPNode>(
            this.BuildNode(split.a),
            this.BuildNode(split.b));

        return node;
    }

    private Pair<ArrayList<Wall>, ArrayList<Wall>> ExecuteSplit(Wall splitWall, ArrayList<Wall> walls)
    {
        DivLine divLine = GetDivLine(splitWall);

        ArrayList<Wall> frontWalls = new ArrayList<Wall>();
        ArrayList<Wall> backWalls = new ArrayList<Wall>();

        for (int i = 0; i < walls.size(); i++)
        {
            Wall w = walls.get(i);
                
            Side side = Side.UNDEFINED;
            if (w.Id == splitWall.Id)
                side = Side.COINCIDENT;
            else
                side = WallOnSide(divLine, w);

            switch(side)
            {
                case COINCIDENT:
                    frontWalls.add(w);
                    backWalls.add(w);
                    break;
                case FRONT:
                    frontWalls.add(w);
                    break;
                case BACK:
                    backWalls.add(w);
                    break;
                case INTERSECTING:

                    // Remove wall from list
                    walls.remove(w);
                    this.Map.Walls.remove(w);

                    // Get intersection
                    Pair<Float, Float> intersection = GetIntersection(divLine, w);

                    // Create vertex at intersection and add it to the map
                    Vertex newVert = new Vertex(this.Map.Vertices.size(), intersection.a, intersection.b);

                    // Add vertex to map
                    this.Map.Vertices.add(newVert);

                    // Create two new walls
                    Wall w1 = new Wall(w.Id,
                                        new Pair<Vertex, Vertex>(w.Line.a, newVert),
                                        w.Color,
                                        w.Type,
                                        w.MinHeight, w.MaxHeight,
                                        w.FloorHeight, w.CeilingHeight);
                    Wall w2 = new Wall(this.GetWallId(),
                                        new Pair<Vertex, Vertex>(newVert, w.Line.b),
                                        w.Color,
                                        w.Type,
                                        w.MinHeight, w.MaxHeight,
                                        w.FloorHeight, w.CeilingHeight);

                    // Add walls to map
                    this.Map.Walls.add(w1);
                    this.Map.Walls.add(w2);

                    // Add walls to front and back lists
                    if (WallOnSide(divLine, w1) == Side.FRONT)
                    {
                        frontWalls.add(w1);
                        backWalls.add(w2);
                    }
                    else
                    {
                        frontWalls.add(w2);
                        backWalls.add(w1);
                    }
                    break;

                default:
                    System.err.println("ERROR::UNKNOWN_SIDE - Unknown side for wall with ID " + w.Id);
                    break;
            }
        }
        return new Pair<ArrayList<Wall>, ArrayList<Wall>>(frontWalls, backWalls);
    }

    private int EvaluateSplit(Wall splitWall, ArrayList<Wall> walls, int bestScore)
    {
        DivLine divLine = GetDivLine(splitWall);

        int frontCount = 0;
        int backCount = 0;

        for (Wall w : walls)
        {
            Side side = Side.UNDEFINED;
            if (w.Id == splitWall.Id)
                continue;
            
            side = WallOnSide(divLine, w);

            // Update the counter
            if (side == Side.FRONT)
                frontCount++;
            else if (side == Side.BACK)
                backCount++;
            else if (side == Side.INTERSECTING)
            {
                frontCount++;
                backCount++;
            }

            // TODO: look into this, this is an optimization that may or may not work.
            // int maxCount = frontCount < backCount ? backCount : frontCount;
            
            // // Get count of walls that need to be split
            // int newCount = frontCount + backCount - walls.size();

            // ??? This comes from DoomBSP, why the 8?
            // score = maxCount + newCount * 8;

            // get out early, i think this may cause some issues in not getting the best
            // score for a given splitwall
            // if (score > bestScore) 
            //     return score;
        }
        if (frontCount == 0 || backCount == 0)
            return MAX_INT;


        int maxCount = frontCount < backCount ? backCount : frontCount;    
        int newCount = frontCount + backCount - walls.size();
        int score = maxCount + newCount * 8;

        return score;
    }
 
    private Side VertOnSide(DivLine divLine, Vertex v)
    {
        // Get vector normal to divLine
        Pair<Float, Float> normalVect = new Pair<Float, Float>(
            -1 * divLine.Direction.b, divLine.Direction.a);

        // Get vector from divLine.Position to v
        Pair<Float, Float> dv = new Pair<Float, Float>(
            v.x - divLine.Position.a, v.y - divLine.Position.b);

        // Get dot product dv * normalVect
        float dot = (dv.a * normalVect.a) + (dv.b * normalVect.b);

        // Return the correct side
        if (dot == 0)
            return Side.INTERSECTING;
        if (dot > 0)
            return Side.FRONT;
        return  Side.BACK;
    }

    private Side WallOnSide(DivLine divLine, Wall w)
    {
        Side v1Side = VertOnSide(divLine, w.Line.a);
        Side v2Side = VertOnSide(divLine, w.Line.b);

        // Coincident
        if (v1Side == Side.INTERSECTING && v2Side == Side.INTERSECTING)
            return Side.COINCIDENT;
        
        // Both on same side
        if (v1Side == v2Side)
            return v1Side;

        // Intersecting at an edge
        if (v1Side == Side.INTERSECTING)
            return v2Side;
        if (v2Side == Side.INTERSECTING)
            return v1Side;

        // Intersecting somewhere in between
        return Side.INTERSECTING;
    }

    private DivLine GetDivLine(Wall w)
    {
        DivLine line = new DivLine();
        line.Position = new Pair<Float, Float>(
            w.Line.a.x, w.Line.a.y);
        line.Direction = new Pair<Float, Float>(
            w.Line.b.x - w.Line.a.x, w.Line.b.y - w.Line.a.y);
        return line;
    }

    private Pair<Wall, Integer> PickSplitWall(ArrayList<Wall> walls)
    {
        int bestScore = MAX_INT;
        Wall bestWall = null;

        for (Wall w : walls)
        {
            int score = EvaluateSplit(w, walls, bestScore);
            if (score < bestScore)
            {
                bestScore = score;
                bestWall = w;
            }
        }
        return new Pair<Wall, Integer>(bestWall, bestScore);
    }

    private Pair<Float, Float> GetIntersection(DivLine divLine, Wall w)
    {

        // Solving Ax = B for x => x = A^-1 B
        Mat2 A = new Mat2(  -1 * divLine.Direction.a, w.Line.b.x - w.Line.a.x,
                            -1 * divLine.Direction.b, w.Line.b.y - w.Line.a.y);

        Vec2 B = new Vec2(divLine.Position.a - w.Line.a.x, divLine.Position.b - w.Line.a.y);

        // If det(A) == 0, lines are coincident, so return v1
        if (A.Det() == 0)
            return null; //new Pair<Float, Float>(w.Line.a.x, w.Line.a.y);

        Vec2 x = Mat2.Mult(A.Inv(), B);
        float t = x.x0;

        // Calculate intersection based on parameterized line
        return new Pair<Float, Float>(divLine.Position.a + (divLine.Direction.a * t),
            divLine.Position.b + (divLine.Direction.b * t));
    }
   
    private int GetWallId()
    {
        if (this.NextWallId == -1)
        {
            int maxId = -1;
            for (Wall w : this.Map.Walls)
            {
                if (w.Id > this.NextWallId)
                    this.NextWallId = w.Id ;
            }
            this.NextWallId = maxId + 1;
        }
        return this.NextWallId++;
    }

    private int GetNodeId()
    {
        return this.NextNodeId++;
    }
}